src/Params/ADnoteParameters.cpp

```diff
@@ -319,6 +351,230 @@ static const Ports voicePorts = {
+                    // so take what we need, bit by bit
+                    if(sem_idx == all_semantics())
+                    {
+                        std::size_t nwaves = waves2->size();
+                        assert(nwaves <= wt->write_space_delayed_semantics(freq_idx));
+                        for(std::size_t sem_idx = 0; sem_idx < nwaves; ++sem_idx)
+                        {
+                            Tensor1<WaveTable::float32>& unusedWave = (*waves2)[sem_idx];
+                            wt->swapDataAt(sem_idx, freq_idx, unusedWave);
+                        }
+                        wt->inc_write_pos_delayed_semantics(freq_idx, nwaves);
+                    } else {
+                        Tensor1<WaveTable::float32>& unusedWave = (*waves1);
+                        wt->swapDataAt(sem_idx, freq_idx, unusedWave);
+                        wt->inc_write_pos_delayed_semantics(freq_idx, 1);
+                    }
+
+                    // recycle the packaging
+                    // this will also free the contained Tensor1 (which are useless after the swap)
+                    if(sem_idx == all_semantics())
+                        d.reply("/free", "sb", "Tensor2<WaveTable::float32>", sizeof(Tensor2<WaveTable::float32>*), &waves2);
+                    else
+                        d.reply("/free", "sb", "Tensor1<WaveTable::float32>", sizeof(Tensor1<WaveTable::float32>*), &waves1);
+
+                    // if this was the last index, the new tensor is complete, so
+                    // we can use it
+                    // -1 for C-style array indexing
+                    if(fillNext && freq_idx == wt->size_freqs() - 1)
+                    {
+#ifdef DBG_WAVETABLES
+                        printf("WT: AD: swap tensors\n");
+#endif
+                        current->swapWith(*wt);
+                        current->set_current_timestamp(paramChangeTime);
+                    }
+                }
+                else
+                {
+                    // since we request the wavetables in order, and the
+                    // communication with MiddleWare is lossless, this should
+                    // imply a programming error
+                    printf("WARNING: MW sent (out-dated?) semantic \"%d\", "
+                           "but the next required semantic is \"%d\" - ignoring!\n",
+                           (int)sem_idx, (int)wt->write_pos_delayed_semantics(freq_idx));
+                    assert(false); // in debug mode, just abort now
+                }
+            }
+        }
+    },
+
+    //Send Messages To Oscillator Realtime Table
+    {"OscilSmp/", rDoc("Primary Oscillator"),
+        &OscilGen::ports,
+        rBOIL_BEGIN
+        if(obj->OscilGn == NULL) return;
+        data.obj = obj->OscilGn;
+        if(strstr(msg, "paste"))
+        {
+            SNIP
+            OscilGen::ports.dispatch(msg, data);
+        }
+        else
+            data.forward();
```

Comment:
ok, this looks sane from the perspective of keeping the normal parameter access done on the right thread. Paste however seems to be done incorrectly? Wouldn't this create data races with a paste operation changing values on the RT thread and wavetable generation which could be occurring on the non-RT thread(s)?

Reply:
I think this relates to [this 3 years old comment](https://github.com/zynaddsubfx/zynaddsubfx/pull/61#issuecomment-731836126) and subsequent discussions which led to [adding a paste handler and letting RT do the paste](https://github.com/zynaddsubfx/zynaddsubfx/commit/c2b41a9f8fd90c4d3197804b14f2417926349535). To cite the commit's comment:

```
Do OscilGen paste on RT side (again)
Let the RT side do the OscilGen paste again, like it is on master. However,
this time, MiddleWare owns the OscilGen, so this commit lets it keep track of
any paste in progress, in order to avoid accessing OscilGen while pasting onto
it.

All changes:

1. New paste state handler, to forbid messages to OscilGen which is currently
   pasted onto
2. Let MW forward OscilGen paste messages to RT
3. Let ADnoteParameters dispatch OscilGen paste messages to OscilGen on the RT
   side (previously, all messages have been forwarded to the front-end)
4. Let OscilGen reply with an `/rt_paste_done` message
5. Let MW delay the `/damage` replies from the `/presets/paste` ports to when
   `/rt_paste_done` has been received by the backend. This means the old
   `/damage` call in `/presets/` - and the accidentally added `/damage`
   call in `/presets/paste:s` - are now deleted.
```

However, 1 year later, I [removed the PasteStateHandler](https://github.com/zynaddsubfx/zynaddsubfx/commit/c1a35909c1615a477ca7b17eddfe98f779663dd1):

```
Handle all non-RT pastes in non-RT
```

for some weird reason, but I forgot to change the callback which you marked here. This commit is also contradicting "Pasting must be realtime" in our discussion. I am not sure yet why I added it, probably some data race? I have to re-think this.